### PR TITLE
Container development guide

### PR DESCRIPTION
**Pull request by brianhlin**
I'm hoping that having a public PR will shame me into actually getting this done...

### CODE REVIEW

File: docs/software/container-development-guide.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+1. Create a `Dockerfile` based off of the OSG Software Base image:
+
+        FROM opensciencegrid/software-base:fresh
+
+        LABEL maintainer OSG Software <help@opensciencegrid.org>
+        
+        RUN yum clean all && \
+            yum update -y 
+
+        RUN yum install -y <PACKAGE> --enablerepo=osg-development
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
        RUN yum install -y <PACKAGE> --enablerepo=osg-development && rm -rf /var/cache/yum/*
```

File: docs/software/container-development-guide.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        RUN yum clean all --enablerepo=* && rm -rf /var/cache/yum/
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
```

File: docs/software/container-development-guide.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+1. Create a `README.md` file describing the software provided by the image
+1. Create a `LICENSE` file containing the [Apache 2.0 license text](https://www.apache.org/licenses/LICENSE-2.0.txt)
+1. Create a `Dockerfile` based off of the OSG Software Base image:
+
+        FROM opensciencegrid/software-base:fresh
+
+        LABEL maintainer OSG Software <help@opensciencegrid.org>
+        
+        RUN yum clean all && \
+            yum update -y
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
            yum update -y && \
            rm -rf /var/cache/yum/*
```

File: docs/software/container-development-guide.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+1. Create a Git repository whose name is prefixed with `docker-`, e.g. `docker-frontier-squid`
+1. Create a `README.md` file describing the software provided by the image
+1. Create a `LICENSE` file containing the [Apache 2.0 license text](https://www.apache.org/licenses/LICENSE-2.0.txt)
+1. Create a `Dockerfile` based off of the OSG Software Base image:
+
+        FROM opensciencegrid/software-base:fresh
+
+        LABEL maintainer OSG Software <help@opensciencegrid.org>
+        
+        RUN yum clean all && \
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
        RUN \
```
software-base is already clean so the initial yum clean all isn't necessary.

File: docs/software/container-development-guide.md (hunk 3 above)
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
            yum update -y && yum clean all && rm -rf /var/cache/yum/*
```
on the other hand, you do need to clean up afterward.

File: docs/software/container-development-guide.md (hunk 2 above)
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
```
this doesn't help.

File: docs/software/container-development-guide.md (hunk 3 above)
Comment (matyasselmeci) [reply docs/software/container-development-guide.md]: You should combine the yum update and the subsequent yum install into one RUN command.

File: docs/software/container-development-guide.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+### Built-in integration ###
+
+Managing Tags in DockerHub
+--------------------------
+
+### Adding tags ###
+
+Images that have passed acceptance testing should be tagged as `stable`:
+
+1. Install the `jq` utility
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
1. Install the `jq` utility:

        yum install jq
        
```

**Comment by brianhlin** [comment]
FYI: @efajardo @edquist @timtheisen @djw8605 

The tag removal bit could be improved since Carl has an automatic clean-up tool now but this is good enough for publishing now

File: docs/software/container-development-guide.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    Replacing `<PACKAGE>` with the name of the RPM you'd like to provide in this container image
+
+### Prepare the DockerHub repository ###
+
+1. Create a Docker Hub repo in the OSG organization.
+1. Go to the permissions tab and give the `robots` team `Read & Write` access
+1. Copy over `.travis.yml` and `travis/` from a previous docker build (e.g., <https://github.com/opensciencegrid/docker-frontier-squid>).
+   Update the `docker_repos` in `build_docker.sh` to the name of the Docker Hub repo
+1. Set the `+x` bit on `travis/build_docker.sh`
+1. Send the `opensciencegrid` GitHub repoository URL to the Software Manager and ask them to do the following:
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
1. Send the `opensciencegrid` GitHub repository URL to the Software Manager and ask them to do the following:
```

File: docs/software/container-development-guide.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  `master`, e.g. <https://travis-ci.org/opensciencegrid/docker-xcache>, and click "Restart build"
+- **If changes need to be made to the container packaging:** submit a pull request with your changes to the relevant
+  GitHub repository and request that another team member review it.
+  Once merged into `master`, a Travis-CI build should start automatically.
+
+If the Travis-CI build completes successfully, you should shortly see new `fresh` and timestamp tags appear in the
+DockerHub repository.
+
+!!! note "Automatic weekly rebuilds"
+    If the repo's Travis-CI is configured as above, container images will automatically rebuilds, and therefore pick up
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]:
```suggestion
    If the repo's Travis-CI is configured as above, container images will automatically rebuild, and therefore pick up
```

**Comment by edquist** [comment]
> FYI: @efajardo @edquist @timtheisen @djw8605
> 
> The tag removal bit could be improved since Carl has an automatic clean-up tool now but this is good enough for publishing now

For future reference:

https://github.com/opensciencegrid/release-tools/blob/master/dockerhub-prune-tags.py

File: docs/software/container-development-guide.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    new packages available in minefield once per week.
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]: Not sure about this note as a whole. The rebuilds happen once a week on a random day of the week so if you're trying to do software development, you shouldn't just wait for a week for the rebuild to happen.

File: docs/software/container-development-guide.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+1. Run the Docker container image tagging command from [release-tools](https://github.com/opensciencegrid/release-tools/):
+
+        ./dockerhub-tag-fresh-to-stable.sh <IMAGE NAME> <TIMESTAMP TAG>
+
+1. Clean up your Docker Hub user/pass file
+
+### Removing tags ###
+
+Via the web interface, requires write permissions
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]: Any details here? Or is it pretty obvious how to do it once you get there?

File: docs/software/container-development-guide.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+Making Slim Containers
+----------------------
+
+Here are some resources for creating slim, efficient containers:
+
+- <https://github.com/opensciencegrid/topology/pull/399>
+- <https://docs.docker.com/develop/develop-images/multistage-build/>
```
Comment (matyasselmeci) [inline docs/software/container-development-guide.md]: I don't think you need three links for multi-stage builds. It's cool but it's really only good for when stage 1 is compiling your software and stage 2 is installing it. Most of our containers are 'yum install stuff from the osg repos and do some config' which doesn't benefit from multi-stage builds.

This link might be useful: https://developers.redhat.com/blog/2016/03/09/more-about-docker-images-size/

File: docs/software/container-development-guide.md (hunk 8 above)
Comment (brianhlin) [reply docs/software/container-development-guide.md]: I think I want to keep this knowledge here so I just changed the admonition to an "info"

File: docs/software/container-development-guide.md (hunk 9 above)
Comment (brianhlin) [reply docs/software/container-development-guide.md]: Wrote a short sentence on how to use Carl's pruner tool

**Pull request merged**